### PR TITLE
Feat/init overwrite

### PR DESCRIPTION
**Pull request by BlaineHeffron**
### What

Adds an --overwrite option to `contract init`

### Why

Useful if you don't want to delete the files yourself - requested here https://github.com/stellar/stellar-cli/issues/1425

### Known limitations

N/A

### CODE REVIEW

File: cmd/soroban-cli/src/commands/contract/init.rs [hunk 1]
```diff
@@ -142,22 +151,25 @@ fn init(
...
 
     Ok(())
 }
 
-fn copy_template_files(project_path: &Path) -> Result<(), Error> {
+fn copy_template_files(project_path: &Path, overwrite: bool) -> Result<(), Error> {
     for item in TemplateFiles::iter() {
         let mut to = project_path.join(item.as_ref());
-
+        let mut exists = false;
```
Comment (willemneal) [inline cmd/soroban-cli/src/commands/contract/init.rs:163]:
```suggestion
        let exists = file_exists(&to);
```

File: cmd/soroban-cli/src/commands/contract/init.rs [hunk 2]
```diff
@@ -142,22 +151,25 @@ fn init(
...
         if file_exists(&to) {
```
Comment (willemneal) [inline cmd/soroban-cli/src/commands/contract/init.rs:164]:
```suggestion
        if exists && !overwrite {
```

File: cmd/soroban-cli/src/commands/contract/init.rs [hunk 3]
```diff
@@ -142,22 +151,25 @@ fn init(
...
-            );
-            continue;
+            exists = true;
+            if !overwrite {
+                println!(
+                    "ℹ️  Skipped creating {} as it already exists",
+                    &to.to_string_lossy()
+                );
+                continue;
+            }
```
Comment (willemneal) [inline cmd/soroban-cli/src/commands/contract/init.rs:172]:
```suggestion
            println!(
                "ℹ️  Skipped creating {} as it already exists",
                &to.to_string_lossy()
            );
            continue;
```

File: cmd/soroban-cli/src/commands/contract/init.rs [hunk 4]
```diff
@@ -223,24 +239,34 @@ fn copy_contents(from: &Path, to: &Path) -> Result<(), Error> {
                 eprintln!("Error creating directory: {new_path:?}");
                 e
             })?;
-            copy_contents(&path, &new_path)?;
+            copy_contents(&path, &new_path, overwrite)?;
         } else {
+            let mut exists = false;
```
Comment (willemneal) [inline cmd/soroban-cli/src/commands/contract/init.rs:244]: Same thing here as above.

File: cmd/soroban-cli/src/commands/contract/init.rs [hunk 5]
```diff
@@ -223,24 +236,32 @@ fn copy_contents(from: &Path, to: &Path) -> Result<(), Error> {
...
+            if exists {
+                let mut appended = false;
                 if new_path.to_string_lossy().contains(".gitignore") {
+                    appended = true;
                     append_contents(&path, &new_path)?;
                 }
                 if new_path.to_string_lossy().contains("README.md") {
+                    appended = true;
                     append_contents(&path, &new_path)?;
                 }
```
Comment (willemneal) [inline cmd/soroban-cli/src/commands/contract/init.rs:251]:
```suggestion
                let new_path_str = new_path.to_string_lossy();
                let append = new_path_str.contains(".gitignore") || new_path_str.contains("README.md");
                if append {
                    append_contents(&path, &new_path)?;
                }
```

File: cmd/soroban-cli/src/commands/contract/init.rs [hunk 6]
```diff
@@ -223,24 +236,32 @@ fn copy_contents(from: &Path, to: &Path) -> Result<(), Error> {
...
+                    println!("🔄  Overwriting {}", &new_path.to_string_lossy());
+                } else {
+                    println!(
+                        "ℹ️  Skipped creating {} as it already exists",
+                        &new_path.to_string_lossy()
+                    );
+                    continue;
+                }
+            } else {
+                println!("➕  Writing {}", &new_path.to_string_lossy());
```
Comment (willemneal) [inline cmd/soroban-cli/src/commands/contract/init.rs:263]:
```suggestion
                if overwrite && !append {
                    println!("🔄  Overwriting {new_path_str}");
                } else {
                    println!(
                        "ℹ️  Skipped creating {new_path_str} as it already exists"
                    );
                    continue;
                }
            } else {
                println!("➕  Writing {new_path_str}");
```

**Review by willemneal — APPROVED** [review]
One more small nit.

File: FULL_HELP_DOCS.md [hunk 7]
```diff
@@ -424,6 +424,7 @@ Initialize a Soroban project with an example contract
 * `-f`, `--frontend-template <FRONTEND_TEMPLATE>` — An optional flag to pass in a url for a frontend template repository.
 
   Default value: ``
+* `-o`, `--overwrite` — Overwrite all existing files.
```
Comment (leighmcculloch) [inline FULL_HELP_DOCS.md:427]: I don't think we should add the short option `-o`, opened a PR removing it and why is mentioned there:
- https://github.com/stellar/stellar-cli/pull/1483

File: cmd/soroban-cli/src/commands/contract/init.rs [hunk 8]
```diff
@@ -47,6 +47,9 @@ pub struct Cmd {
         long_help = "An optional flag to pass in a url for a frontend template repository."
     )]
     pub frontend_template: String,
+
+    #[arg(short, long, long_help = "Overwrite all existing files.")]
```
Comment (leighmcculloch) [inline cmd/soroban-cli/src/commands/contract/init.rs:51]: Is the intent that the command will overwrite the entire directory such as remove the directory contents first, then start writing? Or is the intent that the command will overwrite individual files and leave any additional files as they are?

The term 'all' may introduce some ambiguity to the scope of what is removed, and given the destructive nature of this command we should try and limit the ambiguity.

The term 'any' may limit the ambiguity, although not perfectly.

```suggestion
    #[arg(short, long, long_help = "Overwrite any existing files.")]
```

**Pull request merged**